File: python/백준/그래프/프로이드-워셜/1389 케빈 베이컨의 6단계 법칙.py
import sys

input = sys.stdin.readline
N, M = map(int, input().split())

INF = 1e9

graph = [[INF] * (N + 1) for _ in range(N + 1)]

for a in range(1, N + 1):
    for b in range(1, N + 1):
        if a == b:
            graph[a][b] = 0
            
for _ in range(M):
    a, b = map(int, input().split())
    
    graph[a][b] = 1
    graph[b][a] = 1

for k in range(1, N + 1):
    for a in range(1, N + 1):
        for b in range(1, N + 1):
            graph[a][b] = min(graph[a][b], graph[a][k] + graph[k][b])

count_array = [0] * (N + 1)

for a in range(1, N + 1):
    for b in range(1, N + 1):
        if graph[a][b] != INF:
            count_array[a] += graph[a][b]
   
min_value = min(count_array[1:])  # 1번부터 N번까지 최소값 찾기
min_index = count_array.index(min_value)  # 최소값을 가진 인덱스 찾기
print(min_index)

File: cpp/백준/그래프/프로이드-워셜/1389 케빈 베이컨의 6단계 법칙.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int INF = 1e9;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;

    vector<vector<int>> graph(N + 1, vector<int>(N + 1, INF));
    for (int a = 1; a <= N; ++a) graph[a][a] = 0;

    for (int i = 0; i < M; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a][b] = 1;
        graph[b][a] = 1;
    }

    for (int k = 1; k <= N; ++k) {
        for (int a = 1; a <= N; ++a) {
            for (int b = 1; b <= N; ++b) {
                graph[a][b] = min(graph[a][b], graph[a][k] + graph[k][b]);
            }
        }
    }

    vector<int> count_array(N + 1, 0);
    for (int a = 1; a <= N; ++a) {
        for (int b = 1; b <= N; ++b) {
            if (graph[a][b] != INF) count_array[a] += graph[a][b];
        }
    }

    int min_value = count_array[1];
    int min_index = 1;
    for (int i = 2; i <= N; ++i) {
        if (count_array[i] < min_value) {
            min_value = count_array[i];
            min_index = i;
        }
    }

    cout << min_index << "\n";
    return 0;
}
